LinqToSqlAdapter.UnitTests/QueryResultRetrieverTest.cs

```diff
@@ -77,68 +78,69 @@ public void GetResults_CreatesCommandAndReadsData ()
 
       var result = retriever.GetResults (_projection, "Text", new CommandParameter[0]).ToArray();
 
       Assert.That (result, Is.Empty);
 
-      _dataReaderMock.AssertWasCalled (mock => mock.Dispose());
-      _commandMock.AssertWasCalled (mock => mock.Dispose());
-      _connectionMock.AssertWasCalled (mock => mock.Dispose());
+      _dataReaderMock.Verify (mock => mock.Dispose());
+      _commandMock.Verify (mock => mock.Dispose());
+      _connectionMock.Verify (mock => mock.Dispose());
     }
 
     [Test]
     public void GetResults_SetsCommandData ()
     {
-      _dataReaderMock.Stub (stub => stub.Read()).Return (false);
+      _dataReaderMock.Setup (stub => stub.Read()).Returns (false);
 
-      var dataParameterCollectionMock = MockRepository.GenerateStrictMock<IDataParameterCollection>();
+      var dataParameterCollectionMock = new Mock<IDataParameterCollection> (MockBehavior.Strict);
       dataParameterCollectionMock
-          .Stub (mock => mock.Add (Arg<IDbDataParameter>.Is.Equal (_dataParameter)))
-          .Return (0);
-      dataParameterCollectionMock.Replay();
-      _commandMock.Stub (stub => stub.Parameters).Return (dataParameterCollectionMock);
+          .Setup (mock => mock.Add (It.Is<IDbDataParameter> (d => d.Equals(_dataParameter.Object))))
+          .Returns (0);
+      _commandMock.SetupGet (stub => stub.Parameters).Returns (dataParameterCollectionMock.Object);
 
-      var retriever = new QueryResultRetriever (_connectionManagerStub, _resolverStub);
+      var retriever = new QueryResultRetriever (_connectionManagerStub.Object, _resolverStub.Object);
 
       var result = retriever.GetResults (_projection, "Text", new[] { new CommandParameter ("p1", "value1") }).ToArray();
 
       Assert.That (result, Is.Empty);
 
-      _dataParameter.AssertWasCalled (mock => mock.ParameterName = "p1");
-      _dataParameter.AssertWasCalled (mock => mock.Value = "value1");
-      _commandMock.AssertWasCalled (mock => mock.CommandText = "Text");
-      dataParameterCollectionMock.VerifyAllExpectations();
+      _dataParameter.VerifySet (mock => mock.ParameterName = "p1");
+      _dataParameter.VerifySet (mock => mock.Value = "value1");
+      _commandMock.VerifySet (mock => mock.CommandText = "Text");
+      dataParameterCollectionMock.Verify();
     }
 
     [Test]
     public void GetResults_UsesProjection ()
     {
-      _dataReaderMock.Stub (stub => stub.Read()).Return (true).Repeat.Once();
-      _dataReaderMock.Stub (stub => stub.GetValue (0)).Return ("testColumnValue1").Repeat.Once();
-      _dataReaderMock.Stub (stub => stub.Read()).Return (false);
+      var sequence = new MockSequence();
+      _dataReaderMock.InSequence (sequence).Setup (stub => stub.Read()).Returns (true);
+      _dataReaderMock.InSequence (sequence).Setup (stub => stub.GetValue (0)).Returns ("testColumnValue1");
+      _dataReaderMock.InSequence (sequence).Setup (stub => stub.Read()).Returns (false);
 
```

Comment:
@dukorama @MichaelKetting @iam-mholle consider the following code:

```csharp
    [Test]
    public void MyTest ()
    {
      void DoTheThing (IMyInterface myInterface)
      {
        myInterface.DoTheThing ("A");
      }

      var myMock = new Mock<IMyInterface> (MockBehavior.Strict);
      myMock.SetupSequence (x => x.DoTheThing ("A")).Returns ("1").Returns ("2");
      DoTheThing (myMock.Object);

      myMock.Verify();
    }
```

I create a strict mock and use `SetupSequence` to set up 2 calls, yet `DoTheThing` only calls it once, still the test **succeeds**, meaning `SetupSequence` is unfit to set up expectations in any case.

To my surprise, using `InSequence` behaved the same way, strict or loose made no difference, the test succeeded:

```csharp
    [Test]
    public void MyTest2 ()
    {
      void DoTheThing (IMyInterface myInterface)
      {
        myInterface.DoTheThing ("A");
      }

      var myMock = new Mock<IMyInterface> (MockBehavior.Strict);
      var sequence = new MockSequence();
      myMock.InSequence (sequence).Setup (x => x.DoTheThing ("A")).Returns ("1");
      myMock.InSequence (sequence).Setup (x => x.DoTheThing ("A")).Returns ("2");
      DoTheThing (myMock.Object);

      myMock.Verify();
    }
```

When a method is called that was not setup, strict and loose behave as we would expect them to, e.g this fails for both `SetupSequence` and `InSequence`:

```csharp
    [Test]
    public void MyTest2 ()
    {
      void DoTheThing (IMyInterface myInterface)
      {
        myInterface.DoTheThing ("A");
        myInterface.DoTheThing ("B");
      }

      var myMock = new Mock<IMyInterface> (MockBehavior.Strict);
      var sequence = new MockSequence();
      myMock.InSequence (sequence).Setup (x => x.DoTheThing ("A")).Returns ("1");
      myMock.InSequence (sequence).Setup (x => x.DoTheThing ("A")).Returns ("2");
      DoTheThing (myMock.Object);

      myMock.Verify();
    }
```

My opinion is still intact, we should use `SetupSequence` for stub-type objects that are supposed to return different objects on repeated calls without setting up expectations around that, since switching the order of method calls does not make the test fail when using `SetupSequence`:

```csharp
    [Test]
    public void MyTest ()
    {
      void DoTheThing (IMyInterface myInterface)
      {
        myInterface.DoTheThing ("B");
        myInterface.DoTheThing ("A");
      }

      var myMock = new Mock<IMyInterface> (MockBehavior.Strict);
      myMock.SetupSequence (x => x.DoTheThing ("A")).Returns ("1").Returns ("2");
      myMock.SetupSequence (x => x.DoTheThing ("B")).Returns ("1").Returns ("2");
      DoTheThing (myMock.Object);

      myMock.Verify();
    }
```

Reply:
Okay, I've found some answered to my questions:
https://github.com/MichaelKetting/MoqTest

In Moq there is no way to ensure that all expected calls have been performed. If we use InSequence, only the trailing calls can be left out, otherwise, it's a free-for-all. This is only relevant when calls do not have a return value we depend on. 

@Stefan-Ilic is there a ticket or documentation for this in Moq?

Reply:
Okay, did further tests: we need to use VerifyNoOtherCalls() instead of Verify() when using a Strict mock if we want to cover not-executed calls as well. Which we do.

Reply:
For continuity: check out my PR for your MoqTest project